File: backend/app/schemas/detection.py
"""
Detection and tracking Pydantic schemas
"""
from pydantic import BaseModel
from typing import List, Dict, Any, Optional
from datetime import datetime


class BoundingBox(BaseModel):
    """Bounding box coordinates"""
    x1: float
    y1: float
    x2: float
    y2: float


class PersonTracking(BaseModel):
    """Individual person tracking result"""
    bbox: BoundingBox
    confidence: float
    class_name: str = "person"
    track_id: Optional[int] = None


class ZoneStatus(BaseModel):
    """Zone status and occupancy"""
    zone_id: int
    workstation_id: int
    person_count: int
    status: str  # work, idle, other
    track_ids: List[int]
    timestamp: datetime


class ZoneAnalysis(BaseModel):
    """Complete zone analysis results"""
    analysis_timestamp: datetime
    zones: Dict[int, ZoneStatus]
    total_persons_detected: int


class DetectionResponse(BaseModel):
    """Response for detection API endpoints"""
    detection_id: int
    workstation_id: int
    timestamp: datetime
    person_count: int
    trackings: List[Dict[str, Any]]  # Raw tracking results
    zone_analysis: Dict[str, Any]    # Zone analysis results
    processing_time_ms: float


class EfficiencyData(BaseModel):
    """Zone efficiency metrics"""
    zone_id: int
    time_window_hours: int
    work_time_minutes: float
    idle_time_minutes: float
    other_time_minutes: float
    efficiency_percentage: float


class StatusChange(BaseModel):
    """Zone status change record"""
    status: str
    timestamp: datetime
    person_count: int


class TrackingAnalysisResponse(BaseModel):
    """Response for zone tracking analysis"""
    zone_id: int
    time_window_hours: int
    efficiency_data: EfficiencyData
    status_history: List[StatusChange]


class TrackingMovement(BaseModel):
    """Person movement record"""
    zone_id: int
    timestamp: datetime
    entry_type: str


class TrackingHistoryResponse(BaseModel):
    """Response for tracking history"""
    track_id: int
    workstation_id: int
    movement_history: List[TrackingMovement]
    total_movements: int


class TrackingSessionData(BaseModel):
    """Tracking session information"""
    track_id: int
    workstation_id: int
    first_seen: datetime
    last_seen: datetime
    total_detections: int
    current_zone_id: Optional[int]
    zones_visited: Optional[List[int]]
    is_active: bool

File: src/types/detection.ts
/**
 * Detection and tracking types for YOLOv11 integration
 */

export interface BoundingBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface Detection {
  bbox: BoundingBox;
  confidence: number;
  class: string;
  track_id: number | null;
}

export interface ZoneAnalysis {
  zone_id: number;
  zone_name: string;
  person_count: number;
  status: 'idle' | 'work' | 'other';
  track_ids: number[];
  timestamp: string;
  rectangle: {
    x_min: number;
    y_min: number;
    x_max: number;
    y_max: number;
  };
}

export interface DetectionResponse {
  detection_id: number;
  workstation_id: number;
  timestamp: string;
  person_count: number;
  trackings: Detection[];
  zone_analysis: {
    stream_id: string;
    analysis_timestamp: string;
    zones: Record<string, ZoneAnalysis>;
    total_persons_detected: number;
  };
  processing_time_ms: number;
  frame_width?: number;  // Original captured frame width for coordinate scaling
  frame_height?: number; // Original captured frame height for coordinate scaling
}

export interface DetectionRequest {
  workstation_id: number;
  frame: Blob;
  confidence_threshold?: number;
  persist_tracking?: boolean;
  frame_width?: number;  // For backend to include in response
  frame_height?: number; // For backend to include in response
}